session.go

```diff
@@ -959,3 +984,16 @@ func (rc remoteCallbackHandler) OnUpdate(_ *proto.Update, respond tunnel_client.
 		}
 	}
 }
+
+func (rc remoteCallbackHandler) OnStopTunnel(stopTunnel *proto.StopTunnel, respond tunnel_client.HandlerRespFunc) error {
+	err := &ngrokError{Message: stopTunnel.Message, ErrCode: stopTunnel.ErrorCode}
+	if rc.stopTunnelHandler != nil {
+		// allow user to return an error and instigate a reconnect
+		user_err := rc.stopTunnelHandler(context.TODO(), rc.sess, stopTunnel.ClientID, err)
+		if user_err != nil {
+			return user_err
+		}
```

Comment:
OK, having looked at the agent side of this, I'm not sure this is really what we want.

As I understand it, this would initiate a "hot" reconnect where the session sends its cookie in the auth request and tries to rebind all of its tunnels with the hostports that it was given the last time they were bound. Since we expect this to only really occur (right now at any rate) when this agent lost a race to get an ephemeral port assignment, that reconnect would just loop on trying to rebind that tunnel that it already lost. This would just throw the agent agent into a reconnect loop until someone comes along and ctrl+c's it.

We can either retain the original configuration for the tunnel, try to detect this condition, and attempt a new bind with the same config, or we can skip trying to handle reconnecting the session here altogether and leave it up to the caller to decide what to do with their now-dead tunnel. The former sounds like a whole lot of not-very-fun plumbing to me, plus it would mean that the existing Tunnel object would get a new URL, which feels wrong. It's kinda signaled by this handler if we decided to keep it, but then the user has to maintain the ID mappings for all of their bound tunnels if they want to plumb the "this tunnel now has a new URL" context to wherever. 

So I still think we skip this handler altogether along with any attempt at reconnecting the session here, and let the `Accept` error be the way this gets signaled. Then on the agent side, we can use that error to either trigger a full session restart to get brand new tunnel URLs, or it can attempt to bind a new tunnel with the same config as the dead one.

Reply:
tl;dr: Sure, I'll start switching the agent to a cold reconnect, or rebind with same tunnel-config, depending on which is less complex. That will eliminate the need for `ngrok-go`'s `stopTunnelHandler`.

Aye, you have it right, the agent would go into a "hot" reconnect loop, just like the current closing of the session, under the condition of duplicate tcp address.

For context on why it's implemented this way currently: The goal of the [original issue](https://github.com/ngrok-private/ngrok/issues/22761) was to improve the UX of session close by getting an informative message to the user (and why the original plan was just a status message but was extended to the more specific StopTunnel). Coming up with a plan for changing out hostports on a tunnel-config and broadcasting that out was for a [separate future issue](https://github.com/ngrok-private/ngrok/issues/22597). So agent behavior of hot reconnect was maintained, but with an error message for the user.

But we can make other decisions here too. In all cases we can probably surface an error on the UI that will stick, so the menu of options seem to be:

1. Hot reconnect. If for some reason the other tcp tunnel goes away this will self-heal. Currently implemented.
2. Cold reconnect
    a. If it's an ephemeral tcp address a new one will be provided, which could also update in the UI, but there's no callback to get information that this happened into other software.
    b. If it's a user-specified tcp address this will go into a reconnect loop like option 1, unless the other conflicting tunnel goes away.
3. Tunnel rebind with same config. Ends up basically same as cold reconnect for this case, but might be more complicated to implement.
4. No reconnect, so the agent acts like the SDK's: the session goes on, as do any other tunnels (which are unlikely to exist in agent case). No possibility of self-healing, but the likelihood of that is small anyway.

1 is implemented, you're leaning towards 2 or 3. Between 2 and 3 I'd argue for 2 (cold reconnect) since most agents only have the one tunnel, and that's likely less complex (but we'll see).

Between 1 (hot) and 2 (cold) depends on if it's ok to swap out a hostport under someone. I didn't think that was an option at this time, but it sounds like it could be. I'm not passionate either way, this is a rare condition, and we'll make an error message surface regardless, so I'll start pursuing 2/3.

CC'ing product in case there are stronger concerns there: @nijikokun, @salilsub

Reply:
This is removed, the agent will handle this in a different way, as suggested.